src/sql/workbench/parts/notebook/common/models/notebookTextFileModel.ts

```diff
@@ -0,0 +1,256 @@
+import { TextFileEditorModel } from 'vs/workbench/services/textfile/common/textFileEditorModel';
+import { FindMatch } from 'vs/editor/common/model';
+import { NotebookContentChange, INotebookModel } from 'sql/workbench/parts/notebook/common/models/modelInterfaces';
+import { NotebookChangeType } from 'sql/workbench/parts/notebook/common/models/contracts';
+
+export class NotebookTextFileModel {
+	// save active cell's line/column in editor model for the beginning of the source property
+	private sourceBeginRange: Range;
+	// save active cell's line/column in editor model for the beginning of the output property
+	private outputBeginRange: Range;
+	// save active cell guid
+	private activeCellGuid: string;
+
+	constructor() {
+	}
+
+	public get ActiveCellGuid(): string {
+		return this.activeCellGuid;
+	}
+
+	public set ActiveCellGuid(guid: string) {
+		if (this.activeCellGuid !== guid) {
+			this.sourceBeginRange = undefined;
+			this.outputBeginRange = undefined;
+			this.activeCellGuid = guid;
+		}
+	}
+
+	public transformAndApplyEditForSourceUpdate(contentChange: NotebookContentChange, textEditorModel: TextFileEditorModel | UntitledEditorModel): boolean {
+		let cellGuidRange = this.getCellNodeByGuid(textEditorModel, contentChange.cells[0].cellGuid);
+
+		// convert the range to leverage offsets in the json
+		if (contentChange && contentChange.modelContentChangedEvent && this.areRangePropertiesPopulated(cellGuidRange)) {
+			contentChange.modelContentChangedEvent.changes.forEach(change => {
+				let convertedRange: IRange = {
+					startLineNumber: change.range.startLineNumber + cellGuidRange.startLineNumber - 1,
+					endLineNumber: change.range.endLineNumber + cellGuidRange.startLineNumber - 1,
+					startColumn: change.range.startColumn + cellGuidRange.startColumn,
+					endColumn: change.range.endColumn + cellGuidRange.startColumn
+				};
+				// Need to subtract one because we're going from 1-based to 0-based
+				let startSpaces: string = ' '.repeat(cellGuidRange.startColumn - 1);
+				// The text here transforms a string from 'This is a string\n this is another string' to:
+				//     This is a string
+				//     this is another string
+				textEditorModel.textEditorModel.applyEdits([{
+					range: new Range(convertedRange.startLineNumber, convertedRange.startColumn, convertedRange.endLineNumber, convertedRange.endColumn),
+					text: change.text.split('\n').join('\\n\",\n'.concat(startSpaces).concat('\"'))
+				}]);
+			});
+		} else {
+			return false;
+		}
+		return true;
+	}
+
+	public transformAndApplyEditForOutputUpdate(contentChange: NotebookContentChange, textEditorModel: TextFileEditorModel | UntitledEditorModel): boolean {
+		if (Array.isArray(contentChange.cells[0].outputs) && contentChange.cells[0].outputs.length > 0) {
+			let newOutput = JSON.stringify(contentChange.cells[0].outputs[contentChange.cells[0].outputs.length - 1], undefined, '    ');
+			if (contentChange.cells[0].outputs.length > 1) {
```

Comment:
How many outputs can there be? Not understanding what this logic is doing - it initially grabs just the last output item but then if there's multiple outputs it duplicates that with a , inbetween?

Reply:
In a cell, there can be any arbitrary number of outputs. For example, in the SQL kernel, each message is considered one "output". The table is another "output". These are all concatenated together into the outputs array here.

There's an event that gets fired when outputs are added (they can't be changed). So we're getting the last output that was added, converting it to JSON, and inserting it after the previous last output (if outputs already existed) or inserting it into the empty outputs array.

Because we don't care about perfect indentation for the JSON output object, I'm not doing the same processing as we do for sources where the JSON includes the correct number of initial spaces. The JSON is prettified when it's saved.